src/values/ObjectValue.js

```diff
@@ -111,33 +111,43 @@ export default class ObjectValue extends ConcreteValue {
     "$WeakMapData",
     "$WeakSetData",
   ];
+  static trackedPropertyBindingNames = new Map();
 
   getTrackedPropertyNames(): Array<string> {
     return ObjectValue.trackedPropertyNames;
   }
 
   setupBindings(propertyNames: Array<string>) {
     for (let propName of propertyNames) {
-      let desc = { writeable: true, value: undefined };
-      (this: any)[propName + "_binding"] = {
-        descriptor: desc,
-        object: this,
-        key: propName,
-      };
+      let propBindingName = ObjectValue.trackedPropertyBindingNames.get(propName);
+      invariant(propBindingName !== undefined);
+      (this: any)[propBindingName] = undefined;
     }
   }
 
   static setupTrackedPropertyAccessors(propertyNames: Array<string>) {
     for (let propName of propertyNames) {
+      let propBindingName = ObjectValue.trackedPropertyBindingNames.get(propName);
+      if (propBindingName === undefined)
+        ObjectValue.trackedPropertyBindingNames.set(propName, (propBindingName = propName + "_binding"));
       Object.defineProperty(ObjectValue.prototype, propName, {
         configurable: true,
         get: function() {
-          let binding = this[propName + "_binding"];
-          return binding.descriptor.value;
+          let binding = this[propBindingName];
+          return binding === undefined ? undefined : binding.descriptor.value;
         },
         set: function(v) {
-          invariant(!this.isHavocedObject(), "cannot mutate a havoced object");
-          let binding = this[propName + "_binding"];
+          // this.$IsClassPrototype === undefined is true while the object is being initialized
+          invariant(this.$IsClassPrototype === undefined || !this.isHavocedObject(), "cannot mutate a havoced object");
```

Comment:
It is not clear to me how an object can be in a havoced state while it is being initialized and even less clear that this is at all desirable.

Reply:
_isHavoced itself is being set via the constructor

Reply:
The constructor sets it to false. The invariant allows it to be true. I don't see why you would want that.

Reply:
isHavocedObject must not be called when the object is not initialized. It used to generally allow for being called in an uninitialized state, but the way that was written caused two super expensive binding-property look-ups, that I eliminated by not doing that uninitialized check in isHavocedObject anymore... It has to go somewhere.

Reply:
Ah, so set may be called from the constructor and for such calls, you don't want to call isHavocedObject because it uses a tracked property. Makes sense now. Perhaps a comment will help the future reader.

Reply:
Will add comment.